design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+managed resource that an application operator may dynamically provision, but in
+the likely event that managed resource requires supporting managed resources to
+function usefully the application operator must ask an infrastructure operator
+to provision them.
+
+Furthermore, defining a core set of portable resource claims has begun to limit
+Crossplane. Resource claims are subject to the [lowest common denominator]
+problem of multi-cloud; when a claim may provide configuration inputs that may
+be used to match or provision many kinds of managed resource it may support only
+the settings that apply to _all_ compatible managed resources. This is in part
+why Crossplane defines relatively few resource claim kinds. Meanwhile, it's
+possible that an infrastructure operator deploys Crossplane to provide an
+opinionated abstraction for the application operators in their organisation and
+that said organisation only uses AWS. If this organisation values Crossplane's
+separation of concerns but does not need its portability there is no reason that
+its application operators should be limited to resource claims that are
+constrained by supporting all possible providers.
+
+## Goals
+
+The goals of this proposal are to enable infrastructure operators to publish
+resource classes that represent _groups_ of infrastructure, and to define in
+configuration (not code) how those resource classes relate to resource claims.
+
+It is important that the design put forward:
+
+* Support user defined resource claims with arbitrary configuration parameters.
+* Ensures claim-kind-to-class-kind and claim-kind-to-resource-kind compatibility
+  is discoverable through configuration, not code.
+* Ensures that several interdependent resources submitted to Crossplane
+  simultaneously may eventually resolve to functioning infrastructure.
+* Limits unnecessary API churn, especially to `v1beta1` managed resource APIs
+  which must maintain backward compatibility.
+* Maintains the high-fidelity, composable nature of managed resources.
+* Maintain support for both static and dynamic provisioning.
+* Follow [declarative configuration best practices] and Kubernetes API
+  conventions.
+* Empower the separation of concerns between infrastructure and application
+  owners.
+
+Note that _how_ users will define resource claim kinds is out of scope for this
+design. Defining a new Kubernetes resource kind typically involves authoring a
+[Custom Resource Definition] (CRD). A method to simplify and constrain this
+process is desirable, and may be explored in future.
+
+## Proposal
+
+This document proposes Crossplane introduce two new kinds:
+`ResourceConfiguration` and `ResourceConfigurationClass`. These new kinds would
+exist in the `infrastructure.crossplane.io` API group. Note that in this context
+"configuration" should be interpreted as in the dictionary definition of the
+word:
+
+> an arrangement of elements in a particular form, figure, or combination.
```

Comment:
I have a feeling this name choice is going to be controversial. :)

I considered `ResourceGroup`, but that concept exists in Azure (we already have a `kind: ResourceGroup`) and "a class of resource group" didn't seem to make as much sense as "a class of resource configuration"; I personally feel it's important for us to to stick to the "claim and class" terminology that many Kubernetes users can quickly identify and understand.

I also considered sticking with the `AggregateResource` and `AggregateResourceClass` naming from my previous design. The `ResourceConfiguration` described by this design is identical to my previously proposed `AggregateResource`, but the concept falls apart for `ResourceConfigurationClass`, which does not aggregate other, distinct resource classes.

Reply:
I was more intrigued by the naming choice of `infrastructure.crossplane.io`.  Why infrastructure and not `resources`, `core` or `aggregate`?

Reply:
@displague "core" felt like too generic of a catch-all, "resources" is overloaded, and I didn't like "aggregate" for the reasons described. I didn't consider composite or composition. I prefer infrastructure because these kinds manage kinds that are installed by an "infrastructure stack".

Reply:
I think I like `ResourceSet` more than `ResourceGroup` but less than `ResourceConfiguration`, largely because "a class of resource set" doesn't seem to make as much sense to me as "a class of resource configuration". Splitting hairs though - I don't feel super strongly.

Reply:
Honestly, resource configuration doesn't sound like it represents a group of resource configuration even though the dictionary meaning implies that. I feel like `ResourceSet` explicitly says that there are many resources underneath this without any order but I can see how _a class of resource set_ may not make so much sense.

Reply:
I'm feeling less good about "resource configurations" as a concept name after a few weeks. Some options:

1. `ResourceConfiguration` and `ResourceConfigurationClass`
1. `ResourceFormation` and `ResourceFormationClass`
1. `ResourceCluster` and `ResourceClusterClass`
1. `ResourceArrangement` and `ResourceArrangementClass`
1. `ResourceComposition` and `ResourceCompositionClass`
1. `CompositeResource` and `CompositeResourceClass`

Option three should probably be immediately disregarded due to how overloaded the term "cluster" is in Crossplane. I'm currently feeling best about options 5 and 6. One datapoint that may help with naming will be to consider how these constructs are likely to be used. A `CompositeResource` sounds like it represents "one resource made of many", which feels like it makes sense for the use cases discussed in this document; a GKE cluster with several node pools or a database with an IAM role. `ResourceComposition` feels more appropriate if we imagine these groupings to be more arbitrary - like Terraform modules - such as a VPC, three databases, and a cluster.

Reply:
I've renamed this to `CompositeResource` and `CompositeResourceClass` and added some clarification around the aim under the goals section.